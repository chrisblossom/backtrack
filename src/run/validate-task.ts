import Joi from '@hapi/joi';

function validateTask<T>(task: T) {
	const taskValid = Joi.string()
		.required()
		.min(1)
		.label('task')
		.validate(task);

	if (taskValid.error) {
		throw new Error(taskValid.error.annotate());
	}

	const nodeEnvValid = Joi.string()
		.required()
		.min(1)
		.label('process.env.NODE_ENV')
		.validate(process.env.NODE_ENV);

	if (nodeEnvValid.error) {
		throw new Error(nodeEnvValid.error.annotate());
	}
}

export { validateTask };
