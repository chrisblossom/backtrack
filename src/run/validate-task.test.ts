import { validateTask } from './validate-task';

describe('validateTask', () => {
    test('handles valid task', () => {
        const result = validateTask('dev');

        expect(result).toMatchSnapshot();
    });

    test('handles undefined task', () => {
        try {
            expect.hasAssertions();
            // @ts-ignore
            validateTask();
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('handles empty task', () => {
        try {
            expect.hasAssertions();
            validateTask('');
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('handles bad input', () => {
        try {
            expect.hasAssertions();
            validateTask(0);
        } catch (error) {
            expect(error).toMatchSnapshot();
        }
    });

    test('handles process.env.NODE_ENV not set', () => {
        const NODE_ENV = process.env.NODE_ENV;
        delete process.env.NODE_ENV;

        try {
            expect.hasAssertions();
            validateTask('dev');
        } catch (error) {
            expect(error).toMatchSnapshot();
        }

        process.env.NODE_ENV = NODE_ENV;
    });

    test('handles process.env.NODE_ENV is set', () => {
        const result = validateTask('dev');

        expect(result).toMatchSnapshot();
    });
});
