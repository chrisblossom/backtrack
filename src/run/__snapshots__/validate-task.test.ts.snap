// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateTask handles bad input 1`] = `[Error: "task" must be a string]`;

exports[`validateTask handles empty task 1`] = `[Error: "task" is not allowed to be empty]`;

exports[`validateTask handles process.env.NODE_ENV is set 1`] = `undefined`;

exports[`validateTask handles process.env.NODE_ENV not set 1`] = `[Error: "process.env.NODE_ENV" is required]`;

exports[`validateTask handles undefined task 1`] = `[Error: "task" is required]`;

exports[`validateTask handles valid task 1`] = `undefined`;
